include/seqan3/utility/bloom_filter/bloom_filter.hpp

```diff
@@ -0,0 +1,385 @@
+// Copyright (c) 2006-2020, Knut Reinert & Freie Universität Berlin
+// Copyright (c) 2016-2020, Knut Reinert & MPI für molekulare Genetik
+// This file may be used, modified and/or redistributed under the terms of the 3-clause BSD-License
+// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE.md
+// -----------------------------------------------------------------------------------------------------
+
+/*!\file
+ * \author Tobias Loka <tobias.loka AT hpi.de>
+ * \brief Provides seqan3::bloom_filter.
+ */
+
+#pragma once
+
+#include <seqan3/std/algorithm>
+#include <seqan3/std/bit>
+#include <seqan3/search/dream_index/interleaved_bloom_filter.hpp>
+
+#include <sdsl/bit_vectors.hpp>
+
+#include <seqan3/core/concept/cereal.hpp>
+#include <seqan3/core/detail/strong_type.hpp>
+
+namespace seqan3
+{
+
+/*!\brief The bloom filter. A data structure that efficiently answers set-membership queries.
+ * \tparam data_layout_mode_ Indicates whether the underlying data type is compressed. See seqan3::data_layout.
+ * \implements seqan3::cerealisable
+ *
+ * \details
+ *
+ * ### Bloom Filter (BF)
+ *
+ * The [Bloom Filter](https://en.wikipedia.org/wiki/Bloom_filter) is a probabilistic data structure.
+ * A Bloom Filter can be thought of as a bitvector of length `n` and `h` hash functions and is used to determine set
+ * membership. To insert data, the data is hashed by the `h` hash functions (returning values in `[0, n)`) and the
+ * corresponding `h` positions in the bitvector are set to `1`. To query data, i.e. to determine whether the query
+ * belongs to the set the Bloom Filter was built for, the query is hashed by the same `h` hash functions and the
+ * corresponding positions are checked. If all `h` positions contain a `1`, the query is (probably) in the data set.
+ * Since the Bloom Filter has variable length, the hashing is not bijective, i.e. it may return true for a set
+ * membership query even though the query was never inserted into the Bloom Filter. Note that the Bloom Filter
+ * will always return `true` if the query was inserted, i.e. there may be false positives, but no false negatives.
+ * 
+ * ### Querying
+ * To query the Bloom Filter for a value, call seqan3::bloom_filter::contains(value) which returns
+ * true if the k-mer hash is present in the index, and false if the hash is not present.
+ * The value is a hash value of the k-mer to check membership for.
+ * 
+ * To query the Bloom Filter for a range of values, call seqan3::bloom_filter::count(values) which returns the
+ * number of k-mer hits in the Bloom Filter for the given range of values.
+ * 
+ * Please note the results are based on a heuristic data structure and with a certain probability (depending on 
+ * the selected size of the bit vector) you may receive a false positive result.
+ *
+ * Compared to the interleaved bloom filter implementation, the bloom filter does not make use of
+ * agents to determine and count membership.
```

Comment:
Maybe we can do a `### Differences to the Interleaved Bloom Filter` or something and shortly explain what it is and what the differences are